### PR TITLE
RavenDB-19683 Sharding - Support for queries with distinct. 

### PR DESCRIPTION
**Pull request by arekpalinski**
Refactoring of index read operations so we'll have shard specific read operations which will add shard specific metadata to query results if needed.

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19683/Sharding-Queries-with-distinct-return-duplicates

### Additional description

Support for distinct queries and a lot of refactoring work

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Persistence/IndexReadOperationBase.cs [hunk 1]
```diff
@@ -17,7 +16,7 @@
 
 namespace Raven.Server.Documents.Indexes.Persistence
 {
-    public abstract class IndexReadOperationBase : IndexOperationBase
+    public abstract partial class IndexReadOperationBase : IndexOperationBase
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Persistence/IndexReadOperationBase.cs]: can we remove partial?

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexPersistence.cs [hunk 2]
```diff
@@ -437,10 +437,8 @@ public override IndexWriteOperationBase OpenIndexWriter(Transaction writeTransac
 
         public override IndexReadOperationBase OpenIndexReader(Transaction readTransaction, IndexQueryServerSide query = null)
         {
-            CheckDisposed();
-            CheckInitialized();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexPersistence.cs:441]: this should stay?

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs [hunk 3]
```diff
@@ -981,5 +985,37 @@ public override void Dispose()
             _releaseSearcher?.Dispose();
             _releaseReadTransaction?.Dispose();
         }
+
+        protected struct CreateQueryResultParameters
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs]: readonly struct?

File: src/Raven.Server/Documents/Indexes/Persistence/IndexReadOperationBase.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/IndexReadOperationBase.cs]: yes, it wasn't added intentionally

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexPersistence.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexPersistence.cs:441]: yes, it should stay

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/LuceneIndexReadOperation.cs]: yes

**Review by ppekrol — APPROVED** [review]

**Pull request merged**